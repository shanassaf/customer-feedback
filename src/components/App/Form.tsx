import React from 'react';
import styled from 'styled-components';

const StyledForm = styled.form``;
const StyledInput = styled.input``;
const StyledLabel = styled.label``;
const StyledContainer = styled.div``;
const StyledOption = styled.option``;
const StyledSelect = styled.select``;

export const Form = () => {
	return (
		<StyledForm>
			<StyledContainer>
				<StyledLabel>Name</StyledLabel>
				<StyledInput type="text" />
			</StyledContainer>
			<StyledContainer>
				<StyledLabel>Email</StyledLabel>
				<StyledInput type="text" />
			</StyledContainer>
			<StyledContainer>
				<StyledLabel>Rating</StyledLabel>
				<StyledSelect>
					<StyledOption value="1">1</StyledOption>
					<StyledOption value="2">2</StyledOption>
					<StyledOption value="3">3</StyledOption>
					<StyledOption value="4">4</StyledOption>
					<StyledOption value="5">5</StyledOption>
				</StyledSelect>
			</StyledContainer>
			<StyledContainer>
				<StyledInput type="submit" />
			</StyledContainer>
		</StyledForm>
	);
};
