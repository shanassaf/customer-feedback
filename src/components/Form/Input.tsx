import React from 'react';
import styled from 'styled-components';

interface Props {
	setFeedback: (feedback: Feedback) => void;
	feedback: Feedback;
	name: string;
	pattern?: string;
	type: string;
}

const StyledInput = styled.input`
	padding: 10px 4px;
	cursor: pointer;

	:hover {
		border-color: #1b047c;
	}
`;

export const Input = ({ setFeedback, feedback, name, pattern, type }: Props): JSX.Element => {
	return (
		<StyledInput
			type={type}
			required
			name={name}
			value={feedback[name]}
			pattern={pattern}
			onChange={(ev) => {
				setFeedback({ ...feedback, [ev.target.name]: ev.target.value });
			}}
		/>
	);
};
