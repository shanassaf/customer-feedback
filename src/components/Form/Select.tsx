import React from 'react';
import styled from 'styled-components';

interface Props {
	setFeedback: (feedback: Feedback) => void;
	feedback: Feedback;
}

const StyledSelect = styled.select`
	padding: 10px;
	width: 50px;
	cursor: pointer;

	:hover {
		border-color: #1b047c;
	}
`;
const StyledOption = styled.option``;

export const Select = ({ setFeedback, feedback }: Props): JSX.Element => {
	return (
		<StyledSelect
			required
			name="rating"
			value={feedback.rating}
			onChange={(ev) => {
				setFeedback({ ...feedback, rating: ev.target.value });
			}}
		>
			<StyledOption value="0">0</StyledOption>
			<StyledOption value="1">1</StyledOption>
			<StyledOption value="2">2</StyledOption>
			<StyledOption value="3">3</StyledOption>
			<StyledOption value="4">4</StyledOption>
			<StyledOption value="5">5</StyledOption>
		</StyledSelect>
	);
};
