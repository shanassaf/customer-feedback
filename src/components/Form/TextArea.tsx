import React from 'react';
import styled from 'styled-components';

interface Props {
	setFeedback: (feedback: Feedback) => void;
	feedback: Feedback;
}
const StyledTextArea = styled.textarea`
	padding: 10px 4px;
	cursor: pointer;

	:hover {
		border-color: #1b047c;
	}
`;

export const TextArea = ({ setFeedback, feedback }: Props) => {
	return (
		<StyledTextArea
			required
			name="comment"
			value={feedback.comment}
			onChange={(ev) => {
				setFeedback({ ...feedback, comment: ev.target.value });
			}}
		/>
	);
};
