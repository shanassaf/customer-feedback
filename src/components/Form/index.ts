export {Form} from './Form';
export {Input} from './Input';
export {Select} from './Select';
export {TextArea} from './TextArea';