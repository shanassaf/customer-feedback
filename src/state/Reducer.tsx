const Reducer = (state: State, action: Action): State => {
	switch (action.type) {
		case 'ADD_FEEDBACK':
			return {
				...state,
				feedback: [action.feedback, ...state.feedback],
			};
		default:
			return state;
	}
};

export default Reducer;
