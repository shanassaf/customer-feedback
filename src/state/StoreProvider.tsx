import React, { createContext, useReducer } from 'react';
import Reducer from './Reducer';

const initialState: State = { comments: [] };

const StoreProvider = ({ children }) => {
	const [state, dispatch] = useReducer(Reducer, initialState);
	return <Context.Provider value={{ state, dispatch }}>{children}</Context.Provider>;
};

export const Context = createContext<ContextType>(null);
export default StoreProvider;
