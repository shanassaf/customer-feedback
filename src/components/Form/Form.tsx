import React, { useContext, useState } from 'react';
import styled from 'styled-components';
import { v4 as uuidGenerator } from 'uuid';
import { Context } from '../../state/StoreProvider';
import { Input } from './Input';
import { Select } from './Select';
import { TextArea } from './TextArea';

const StyledTitle = styled.h1`
	font-size: 20px;
`;
const StyledSubTitle = styled.h2`
	font-size: 12px;
`;
const StyledForm = styled.form`
	font-size: 10px;
	margin: 10px 0;
`;
const StyledInputButton = styled.input`
	margin-top: 20px;
	height: 40px;
	width: 100px;
	background-color: white;
	border-radius: 5px;
	border-width: 1px;
	font-size: 12px;

	:hover {
		cursor: pointer;
		border-color: #1b047c;
		background: #c5dadd;
	}
`;

const StyledLabel = styled.label`
	padding: 20px 0 10px;
`;
const StyledContainer = styled.div`
	display: flex;
	flex-direction: column;
	margin: 10px 0;
`;

const initialState = {
	name: '',
	email: '',
	rating: '0',
	comment: '',
	uuid: '',
} as Feedback;

export const Form = (): JSX.Element => {
	const { dispatch } = useContext(Context);
	const [feedback, setFeedback] = useState<Feedback>(initialState);

	const onSubmit = (ev: React.FormEvent): void => {
		ev.preventDefault();
		dispatch({ type: 'ADD_FEEDBACK', feedback: { ...feedback, uuid: uuidGenerator() } });
		setFeedback(initialState);
	};

	return (
		<>
			<StyledTitle>Customer Feedback Form</StyledTitle>
			<StyledSubTitle>Please leave a review of our product here!</StyledSubTitle>
			<StyledForm onSubmit={onSubmit}>
				<StyledContainer>
					<StyledLabel htmlFor="name">Name *</StyledLabel>
					<Input setFeedback={setFeedback} feedback={feedback} name="name" type="text" />
				</StyledContainer>
				<StyledContainer>
					<StyledLabel htmlFor="email">Email *</StyledLabel>
					<Input setFeedback={setFeedback} feedback={feedback} name="email" type="email" />
				</StyledContainer>
				<StyledContainer>
					<StyledLabel htmlFor="comment">Comment *</StyledLabel>
					<TextArea setFeedback={setFeedback} feedback={feedback} />
				</StyledContainer>
				<StyledContainer>
					<StyledLabel htmlFor="rating">Rating *</StyledLabel>
					<Select setFeedback={setFeedback} feedback={feedback} />
				</StyledContainer>
				<StyledContainer>
					<StyledInputButton type="submit" value="Submit" />
				</StyledContainer>
			</StyledForm>
		</>
	);
};
