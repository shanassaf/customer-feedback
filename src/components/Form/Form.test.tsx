import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { Form } from './Form';
import { Context } from '../../state/StoreProvider';

describe('Form', () => {
	const dispatch = jest.fn();
	const state = {
		feedback: [
			{
				name: 'name',
				email: 'email',
				comment: 'text',
				rating: '1',
				uuid: 'uuid',
			},
		],
	} as State;

	test('component renders correctly', () => {
		const { asFragment } = render(
			<Context.Provider value={{ dispatch, state }}>
				<Form />
			</Context.Provider>
		);
		expect(asFragment()).toMatchSnapshot();
	});

	test('component dispatches action on submit', () => {
		const { getByRole } = render(
			<Context.Provider value={{ dispatch, state }}>
				<Form />
			</Context.Provider>
		);
		const form = getByRole('button', { name: /submit/i });
		fireEvent.submit(form);
		expect(dispatch).toHaveBeenCalledTimes(1);
	});
});
