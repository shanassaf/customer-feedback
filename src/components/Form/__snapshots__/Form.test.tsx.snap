// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component renders correctly 1`] = `
<DocumentFragment>
  <h1
    class="sc-eCApGN iEtHMq"
  >
    Customer Feedback Form
  </h1>
  <h2
    class="sc-jSFkmK deQnhz"
  >
    Please leave a review of our product here!
  </h2>
  <form
    class="sc-gKAblj hkzmtT"
  >
    <div
      class="sc-pNWxx bWjldX"
    >
      <label
        class="sc-fujyUd cwiAfG"
        for="name"
      >
        Name *
      </label>
      <input
        class="sc-bdnylx bzTEwn"
        name="name"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="sc-pNWxx bWjldX"
    >
      <label
        class="sc-fujyUd cwiAfG"
        for="email"
      >
        Email *
      </label>
      <input
        class="sc-bdnylx bzTEwn"
        name="email"
        required=""
        type="email"
        value=""
      />
    </div>
    <div
      class="sc-pNWxx bWjldX"
    >
      <label
        class="sc-fujyUd cwiAfG"
        for="comment"
      >
        Comment *
      </label>
      <textarea
        class="sc-hKFyIo drMVRo"
        name="comment"
        required=""
      />
    </div>
    <div
      class="sc-pNWxx bWjldX"
    >
      <label
        class="sc-fujyUd cwiAfG"
        for="rating"
      >
        Rating *
      </label>
      <select
        class="sc-gtssRu dgBQYk"
        name="rating"
        required=""
      >
        <option
          class="sc-dlnjPT"
          value="0"
        >
          0
        </option>
        <option
          class="sc-dlnjPT"
          value="1"
        >
          1
        </option>
        <option
          class="sc-dlnjPT"
          value="2"
        >
          2
        </option>
        <option
          class="sc-dlnjPT"
          value="3"
        >
          3
        </option>
        <option
          class="sc-dlnjPT"
          value="4"
        >
          4
        </option>
        <option
          class="sc-dlnjPT"
          value="5"
        >
          5
        </option>
      </select>
    </div>
    <div
      class="sc-pNWxx bWjldX"
    >
      <input
        class="sc-iCoHVE clHQPA"
        type="submit"
        value="Submit"
      />
    </div>
  </form>
</DocumentFragment>
`;
